xbmc/windowing/osx/WinSystemOSX.mm

```diff
@@ -0,0 +1,1376 @@
+{
+  int ret = 0;
+
+  // Add full screen settings for additional monitors
+  int numDisplays = NSScreen.screens.count;
+
+  for (int disp = 0; disp < numDisplays; disp++)
+  {
+    NSString* name = screenNameForDisplay(GetDisplayID(disp));
+    if (name.UTF8String == dispName)
+    {
+      ret = disp;
+      break;
+    }
+  }
+
+  return ret;
+}
+
+#pragma mark - Display Modes
+
+CFArrayRef GetAllDisplayModes(CGDirectDisplayID display)
+{
+  int value = 1;
+
+  CFNumberRef number = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &value);
+  if (!number)
+  {
+    CLog::Log(LOGERROR, "GetAllDisplayModes - could not create Number!");
+    return NULL;
+  }
+
+  CFStringRef key = kCGDisplayShowDuplicateLowResolutionModes;
+  CFDictionaryRef options = CFDictionaryCreate(kCFAllocatorDefault, (const void**)&key,
+                                               (const void**)&number, 1, NULL, NULL);
+  CFRelease(number);
+
+  if (!options)
+  {
+    CLog::Log(LOGERROR, "GetAllDisplayModes - could not create Dictionary!");
+    return NULL;
+  }
+
+  CFArrayRef displayModes = CGDisplayCopyAllDisplayModes(display, options);
+  CFRelease(options);
+
+  if (!displayModes)
+  {
+    CLog::Log(LOGERROR, "GetAllDisplayModes - no displaymodes found!");
+    return NULL;
+  }
+
+  return displayModes;
+}
+
+// try to find mode that matches the desired size, refreshrate
+// non interlaced, nonstretched, safe for hardware
+CGDisplayModeRef GetMode(int width, int height, double refreshrate, int screenIdx)
+{
+  if (screenIdx >= (signed)[[NSScreen screens] count])
+    return NULL;
+
+  bool stretched;
+  bool interlaced;
+  bool safeForHardware;
+  bool televisionoutput;
+  int w, h, bitsperpixel;
+  double rate;
+  RESOLUTION_INFO res;
+
+  CLog::Log(LOGDEBUG, "GetMode looking for suitable mode with %d x %d @ %f Hz on display %d", width,
+            height, refreshrate, screenIdx);
+
+  CFArrayRef displayModes = GetAllDisplayModes(GetDisplayID(screenIdx));
+
+  if (!displayModes)
+    return NULL;
+
+  for (int i = 0; i < CFArrayGetCount(displayModes); ++i)
+  {
+    CGDisplayModeRef displayMode = (CGDisplayModeRef)CFArrayGetValueAtIndex(displayModes, i);
+    uint32_t flags = CGDisplayModeGetIOFlags(displayMode);
+    stretched = (flags & kDisplayModeStretchedFlag) != 0;
+    interlaced = (flags & kDisplayModeInterlacedFlag) != 0;
+    bitsperpixel = DisplayBitsPerPixelForMode(displayMode);
+    safeForHardware = (flags & kDisplayModeSafetyFlags) != 0;
+    televisionoutput = (flags & kDisplayModeTelevisionFlag) != 0;
+    w = CGDisplayModeGetWidth(displayMode);
+    h = CGDisplayModeGetHeight(displayMode);
+    rate = CGDisplayModeGetRefreshRate(displayMode);
+
+
+    if ((bitsperpixel == 32) && (safeForHardware == true) && (stretched == false) &&
+        (interlaced == false) && (w == width) && (h == height) &&
+        (rate == refreshrate || rate == 0))
+    {
+      CLog::Log(LOGDEBUG, "GetMode found a match!");
+      return displayMode;
```

Comment:
you forgot to release the array:

```suggestion
      CFRelease(displayModes);
      return displayMode;
```

Reply:
Have gone through and checked all the cf*ref usage. I believe everything is released/retained where appropriate now